aea/helpers/dialogue/base.py

```diff
@@ -412,185 +403,361 @@ def last_message(self) -> Optional[Message]:
+                    message.message_id,
+                    message.dialogue_reference,
+                    self.dialogue_label.dialogue_reference,
+                )
+            )
+
+        is_valid_result, validation_message = self.validate_next_message(message)
+
+        if not is_valid_result:
+            raise InvalidDialogueMessage(
+                "Message {} is invalid with respect to this dialogue. Error: {}".format(
+                    message.message_id, validation_message,
+                )
+            )
+
+        if self.is_message_by_self(message):
+            self._outgoing_messages.extend([message])
+        else:
+            self._incoming_messages.extend([message])
 
     def is_belonging_to_dialogue(self, message: Message) -> bool:
         """
         Check if the message is belonging to the dialogue.
 
         :param message: the message
-        :return: Ture if message is part of the dialogue, False otherwise
+        :return: True if message is part of the dialogue, False otherwise
         """
+        opponent = self.counterparty_from_message(message)
         if self.is_self_initiated:
             self_initiated_dialogue_label = DialogueLabel(
-                (message.dialogue_reference[0], Dialogue.OPPONENT_STARTER_REFERENCE),
-                message.counterparty,
+                (
+                    message.dialogue_reference[0],
+                    Dialogue.UNASSIGNED_DIALOGUE_REFERENCE,
+                ),
+                opponent,
                 self.agent_address,
             )
             result = self_initiated_dialogue_label in self.dialogue_labels
         else:
             other_initiated_dialogue_label = DialogueLabel(
-                message.dialogue_reference, message.counterparty, message.counterparty
+                message.dialogue_reference, opponent, opponent,
             )
             result = other_initiated_dialogue_label in self.dialogue_labels
         return result
 
-    def reply(self, target_message: Message, performative, **kwargs) -> Message:
+    def reply(
+        self,
+        performative: Message.Performative,
+        target_message: Optional[Message] = None,
+        **kwargs,
+    ) -> Message:
         """
         Reply to the 'target_message' in this dialogue with a message with 'performative', and contents from kwargs.
 
+        Note if no target_message is provided, the last message in the dialogue will be replied to.
+
         :param target_message: the message to reply to.
         :param performative: the performative of the reply message.
         :param kwargs: the content of the reply message.
 
         :return: the reply message if it was successfully added as a reply, None otherwise.
         """
-        assert (
-            self._message_class is not None
-        ), "No 'message_class' argument was provided to this class on construction."
         assert self.last_message is not None, "Cannot reply in an empty dialogue!"
```

Comment:
```suggestion
        last_message = self.last_message 
        assert last_message  is not None, "Cannot reply in an empty dialogue!"
```

Reply:
By doing an assignment we can avoid multiple calls to `self.last_message` and satisfy mypy (see comments below)

Reply:
I think this is not necessary as after I changed the assertion from `is_empty` to checking explicitly that `last_message is not None`, I forgot to remove `type: ignore` form the lines. 

Though for performance reasons it is cleaner to make the call once and then reuse `last_message` across the method so I'm applying your suggestion.

Reply:
Done